assets/src/edit-story/components/form/numeric.js

```diff
@@ -97,48 +111,124 @@ function Numeric({
+    setInputValue(event.target.value);
+  }, []);
+
+  const validateAndSubmitInput = useCallback(() => {
+    targetFormRef.current = inputRef.current.form;
+
+    const validValue = validateInput(inputValue, {
+      canBeNegative,
+      canBeEmpty,
+      float,
+    });
+
+    setInputValue(validValue ?? value);
+    onChange(validValue ?? value);
+  }, [inputValue, canBeNegative, canBeEmpty, float, value, onChange]);
 
   const handleUpDown = useCallback(
     ({ key, altKey }) => {
       if (isMultiple) {
         return;
       }
 
-      let newValue;
+      const validValue = validateInput(inputValue, {
+        canBeNegative,
+        canBeEmpty,
+        float,
+      });
+
+      // The `|| 0` is to prevent newValue from being set to
+      // empty string by either validValue or value
+      let newValue = (validValue ?? value) || 0;
       const diff = Big(float && altKey ? 0.1 : 1);
       if (key === 'ArrowUp') {
         // Increment value
         newValue =
           typeof max !== 'undefined'
-            ? Math.min(max, Big(value).plus(diff))
-            : Big(value).plus(diff);
+            ? Math.min(max, Big(newValue).plus(diff))
+            : Big(newValue).plus(diff);
       } else if (key === 'ArrowDown') {
         // Decrement value
         newValue =
           typeof min !== 'undefined'
-            ? Math.max(min, Big(value).minus(diff))
-            : Big(value).minus(diff);
+            ? Math.max(min, Big(newValue).minus(diff))
+            : Big(newValue).minus(diff);
+      }
+
+      setInputValue(newValue.toString());
+    },
+    [isMultiple, float, max, inputValue, value, min, canBeNegative, canBeEmpty]
+  );
+
+  const handleKeyDownEffect = useCallback(
+    ({ key, altKey }) => {
+      switch (key) {
+        case 'Escape':
+          // Revert input value and exit input focus without
+          // triggering blur validation
+          setInputValue(value);
+          skipValidationRef.current = true;
+          inputRef.current.blur();
+          break;
+        default:
+          handleUpDown({ key, altKey });
       }
-      onChange(parseFloat(newValue.toString()));
     },
-    [isMultiple, float, onChange, max, value, min]
+    [value, handleUpDown]
   );
 
+  const handleKeyDown = useCallback(
+    (event) => {
+      if (event.key === 'Enter') {
+        validateAndSubmitInput();
+      }
+    },
+    [validateAndSubmitInput]
+  );
+
+  const submitForm = useCallback(() => {
+    if (targetFormRef.current) {
+      targetFormRef.current.dispatchEvent(
+        new window.Event('submit', { cancelable: true })
+      );
+
+      // Resetting to null after dispatch, it'll get
+      // re-assigned on component blur
+      targetFormRef.current = null;
+    }
+  }, []);
+
   useKeyDownEffect(
-    ref,
-    { key: ['up', 'alt+up', 'down', 'alt+down'], editable: true },
-    handleUpDown,
-    [handleUpDown]
+    inputRef,
+    {
+      key: ['up', 'alt+up', 'down', 'alt+down', 'escape'],
```

Comment:
Why are these squashed into a single listener and not just two separate listeners? Then the switch in `handleKeyDownEffect` can go away.

Reply:
I tried separate handlers and 'escape' or 'enter' (in separate) handlers didn't register... but I may have done something wrong so I'll try again.